File: ts-model-api/src/GeneratedLanguage.ts
import type { ILanguage } from "./ILanguage.js";
import { LanguageRegistry } from "./LanguageRegistry.js";
import type { INodeJS } from "./INodeJS.js";
import type { TypedNode } from "./TypedNode.js";
import type { GeneratedConcept } from "./GeneratedConcept.js";

export abstract class GeneratedLanguage implements ILanguage {
  public nodeWrappers: Map<string, (node: INodeJS) => TypedNode> = new Map();

  constructor(readonly name: string) {}

  public abstract getConcepts(): Array<GeneratedConcept>;

  public register() {
    LanguageRegistry.INSTANCE.register(this);
  }

  public unregister() {
    LanguageRegistry.INSTANCE.unregister(this);
  }

  public isRegistered(): boolean {
    return LanguageRegistry.INSTANCE.isRegistered(this);
  }

  public assertRegistered() {
    if (!this.isRegistered())
      throw Error("Language " + this.name + " is not registered");
  }

  public getName(): string {
    return this.name;
  }

  public getUID(): string {
    return this.name;
  }
}

File: model-api-gen-runtime/src/commonMain/kotlin/org/modelix/metamodel/GeneratedLanguage.kt
/*
 * Copyright (c) 2023.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.modelix.metamodel

import org.modelix.model.api.ILanguage

abstract class GeneratedLanguage(private val name: String) : ILanguage {
    fun register() {
        TypedLanguagesRegistry.register(this)
    }

    fun unregister() {
        TypedLanguagesRegistry.unregister(this)
    }

    fun isRegistered(): Boolean {
        return TypedLanguagesRegistry.isRegistered(this)
    }

    fun assertRegistered() {
        if (!isRegistered()) throw IllegalStateException("Language ${getUID()} is not registered")
    }

    override fun getName(): String {
        return name
    }

    override fun getUID(): String {
        return getName()
    }
}
